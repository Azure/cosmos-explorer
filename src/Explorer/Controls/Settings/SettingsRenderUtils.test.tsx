import { shallow } from "enzyme";
import React from "react";
import { IColumn, Text } from "@fluentui/react";
import {
  getAutoPilotV3SpendElement,
  getEstimatedSpendingElement,
  manualToAutoscaleDisclaimerElement,
  ttlWarning,
  indexingPolicynUnsavedWarningMessage,
  updateThroughputBeyondLimitWarningMessage,
  updateThroughputDelayedApplyWarningMessage,
  getThroughputApplyDelayedMessage,
  getThroughputApplyShortDelayMessage,
  getThroughputApplyLongDelayMessage,
  getToolTipContainer,
  conflictResolutionCustomToolTip,
  changeFeedPolicyToolTip,
  conflictResolutionLwwTooltip,
  mongoIndexingPolicyDisclaimer,
  mongoIndexingPolicyAADError,
  mongoIndexTransformationRefreshingMessage,
  renderMongoIndexTransformationRefreshMessage,
  ManualEstimatedSpendingDisplayProps,
  PriceBreakdown,
  getRuPriceBreakdown,
} from "./SettingsRenderUtils";

class SettingsRenderUtilsTestComponent extends React.Component {
  public override render(): JSX.Element {
    const estimatedSpendingColumns: IColumn[] = [
      { key: "costType", name: "", fieldName: "costType", minWidth: 100, maxWidth: 200, isResizable: true },
      { key: "hourly", name: "Hourly", fieldName: "hourly", minWidth: 100, maxWidth: 200, isResizable: true },
      { key: "daily", name: "Daily", fieldName: "daily", minWidth: 100, maxWidth: 200, isResizable: true },
      { key: "monthly", name: "Monthly", fieldName: "monthly", minWidth: 100, maxWidth: 200, isResizable: true },
    ];
    const estimatedSpendingItems: ManualEstimatedSpendingDisplayProps[] = [
      {
        costType: <Text>Current Cost</Text>,
        hourly: <Text>$ 1.02</Text>,
        daily: <Text>$ 24.48</Text>,
        monthly: <Text>$ 744.6</Text>,
      },
    ];
    const priceBreakdown: PriceBreakdown = {
      hourlyPrice: 1.02,
      dailyPrice: 24.48,
      monthlyPrice: 744.6,
      pricePerRu: 0.00051,
      currency: "RMB",
      currencySign: "Â¥",
    };

    return (
      <>
        {getAutoPilotV3SpendElement(1000, false)}
        {getAutoPilotV3SpendElement(undefined, false)}
        {getAutoPilotV3SpendElement(1000, true)}
        {getAutoPilotV3SpendElement(undefined, true)}

        {getEstimatedSpendingElement(estimatedSpendingColumns, estimatedSpendingItems, 1000, 2, priceBreakdown, false)}

        {manualToAutoscaleDisclaimerElement}
        {ttlWarning}
        {indexingPolicynUnsavedWarningMessage}
        {updateThroughputBeyondLimitWarningMessage}
        {updateThroughputDelayedApplyWarningMessage}

        {getThroughputApplyDelayedMessage(false, 1000, "RU/s", "sampleDb", "sampleCollection", 2000)}
        {getThroughputApplyShortDelayMessage(false, 1000, "RU/s", "sampleDb", "sampleCollection")}
        {getThroughputApplyLongDelayMessage(false, 1000, "RU/s", "sampleDb", "sampleCollection", 2000)}

        {getToolTipContainer(<span>Sample Text</span>)}
        {conflictResolutionLwwTooltip}
        {conflictResolutionCustomToolTip}
        {changeFeedPolicyToolTip}

        {mongoIndexingPolicyDisclaimer}
        {mongoIndexingPolicyAADError}
        {mongoIndexTransformationRefreshingMessage}
        {renderMongoIndexTransformationRefreshMessage(0, () => {
          return;
        })}
        {renderMongoIndexTransformationRefreshMessage(90, () => {
          return;
        })}
      </>
    );
  }
}

describe("SettingsUtils functions", () => {
  it("render", () => {
    const wrapper = shallow(<SettingsRenderUtilsTestComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should return correct price breakdown for a manual RU setting of 500, 1 region, multimaster disabled", () => {
    const prices = getRuPriceBreakdown(500, "", 1, false, false);
    expect(prices.hourlyPrice).toBe(0.04);
    expect(prices.dailyPrice).toBe(0.96);
    expect(prices.monthlyPrice).toBe(29.2);
    expect(prices.pricePerRu).toBe(0.00008);
    expect(prices.currency).toBe("USD");
    expect(prices.currencySign).toBe("$");
  });
});
