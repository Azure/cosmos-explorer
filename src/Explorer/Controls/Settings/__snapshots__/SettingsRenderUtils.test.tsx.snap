// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsUtils functions render 1`] = `
<Fragment>
  <Text>
    Your 
    container
     throughput will automatically scale from
     
    <b>
      100
       RU/s (10% of max RU/s) -
       
      1000
       RU/s
    </b>
     
    based on usage.
    <br />
  </Text>
  <Text>
    After the first 
    10
     GB of data stored, the max RU/s will be automatically upgraded based on the new storage value.
    <StyledLinkBase
      href="https://aka.ms/cosmos-autoscale-info"
      target="_blank"
    >
       
      Learn more
    </StyledLinkBase>
    .
  </Text>
  <Text>
    Your 
    database
     throughput will automatically scale from
     
    <b>
      100
       RU/s (10% of max RU/s) -
       
      1000
       RU/s
    </b>
     
    based on usage.
    <br />
  </Text>
  <Text>
    After the first 
    10
     GB of data stored, the max RU/s will be automatically upgraded based on the new storage value.
    <StyledLinkBase
      href="https://aka.ms/cosmos-autoscale-info"
      target="_blank"
    >
       
      Learn more
    </StyledLinkBase>
    .
  </Text>
  <Stack
    styles={
      Object {
        "root": Object {
          "width": 600,
        },
      }
    }
    tokens={
      Object {
        "childrenGap": 10,
      }
    }
  >
    <StyledWithViewportComponent
      columns={
        Array [
          Object {
            "fieldName": "costType",
            "isResizable": true,
            "key": "costType",
            "maxWidth": 200,
            "minWidth": 100,
            "name": "",
          },
          Object {
            "fieldName": "hourly",
            "isResizable": true,
            "key": "hourly",
            "maxWidth": 200,
            "minWidth": 100,
            "name": "Hourly",
          },
          Object {
            "fieldName": "daily",
            "isResizable": true,
            "key": "daily",
            "maxWidth": 200,
            "minWidth": 100,
            "name": "Daily",
          },
          Object {
            "fieldName": "monthly",
            "isResizable": true,
            "key": "monthly",
            "maxWidth": 200,
            "minWidth": 100,
            "name": "Monthly",
          },
        ]
      }
      disableSelectionZone={true}
      items={
        Array [
          Object {
            "costType": <Text>
              Current Cost
            </Text>,
            "daily": <Text>
              $ 24.48
            </Text>,
            "hourly": <Text>
              $ 1.02
            </Text>,
            "monthly": <Text>
              $ 744.6
            </Text>,
          },
        ]
      }
      layoutMode={1}
      onRenderRow={[Function]}
      selectionMode={0}
    />
    <Text
      id="throughputSpendElement"
    >
      (
      regions: 
       
      2
      , 
      1000
       RU/s, 
      Â¥
      0.00051
      /RU)
    </Text>
    <Text>
      <em>
        This cost is an estimate and may vary based on the regions where your account is deployed and potential discounts applied to your account
      </em>
    </Text>
  </Stack>
  <Text
    id="manualToAutoscaleDisclaimerElement"
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    The starting autoscale max RU/s will be determined by the system, based on the current manual throughput settings and storage of your resource. After autoscale has been enabled, you can change the max RU/s.
     
    <StyledLinkBase
      href="https://aka.ms/cosmos-autoscale-migration"
    >
      Learn more
    </StyledLinkBase>
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    The system will automatically delete items based on the TTL value (in seconds) you provide, without needing a delete operation explicitly issued by a client application. For more information see,
     
    <StyledLinkBase
      href="https://aka.ms/cosmos-db-ttl"
      target="_blank"
    >
      Time to Live (TTL) in Azure Cosmos DB
    </StyledLinkBase>
    .
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    You have not saved the latest changes made to your indexing policy. Please click save to confirm the changes.
  </Text>
  <Text
    id="updateThroughputBeyondLimitWarningMessage"
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    You are about to request an increase in throughput beyond the pre-allocated capacity. The service will scale out and increase throughput for the selected container. This operation will take 1-3 business days to complete. You can track the status of this request in Notifications.
  </Text>
  <Text
    id="updateThroughputDelayedApplyWarningMessage"
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    You are about to request an increase in throughput beyond the pre-allocated capacity. This operation will take some time to complete.
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    The request to increase the throughput has successfully been submitted. This operation will take 1-3 business days to complete. View the latest status in Notifications.
    <br />
    Database: 
    sampleDb
    , Container: 
    sampleCollection
     
    , Current manual throughput: 1000 RU/s, Target manual throughput: 2000
  </Text>
  <Text
    id="throughputApplyShortDelayMessage"
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    A request to increase the throughput is currently in progress. This operation will take some time to complete.
    <br />
    Database: sampleDb, Container: sampleCollection 
    , Current manual throughput: 1000 RU/s
  </Text>
  <Text
    id="throughputApplyLongDelayMessage"
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    A request to increase the throughput is currently in progress. This operation will take 1-3 business days to complete. View the latest status in Notifications.
    <br />
    Database: sampleDb, Container: sampleCollection 
    , Current manual throughput: 1000 RU/s, Target manual throughput: 2000
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    <span>
      Sample Text
    </span>
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    Gets or sets the name of a integer property in your documents which is used for the Last Write Wins (LWW) based conflict resolution scheme. By default, the system uses the system defined timestamp property, _ts to decide the winner for the conflicting versions of the document. Specify your own integer property if you want to override the default timestamp based conflict resolution.
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    Gets or sets the name of a stored procedure (aka merge procedure) for resolving the conflicts. You can write application defined logic to determine the winner of the conflicting versions of a document. The stored procedure will get executed transactionally, exactly once, on the server side. If you do not provide a stored procedure, the conflicts will be populated in the
    <StyledLinkBase
      className="linkDarkBackground"
      href="https://aka.ms/dataexplorerconflics"
      target="_blank"
    >
       conflicts feed
    </StyledLinkBase>
    . You can update/re-register the stored procedure at any time.
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    Enable change feed log retention policy to retain last 10 minutes of history for items in the container by default. To support this, the request unit (RU) charge for this container will be multiplied by a factor of two for writes. Reads are unaffected.
  </Text>
  <Text>
    For queries that filter on multiple properties, create multiple single field indexes instead of a compound index.
    <StyledLinkBase
      href="https://docs.microsoft.com/azure/cosmos-db/mongodb-indexing#index-types"
      target="_blank"
    >
       Compound indexes 
    </StyledLinkBase>
    are only used for sorting query results. If you need to add a compound index, you can create one using the Mongo shell.
  </Text>
  <StyledMessageBar
    messageBarType={1}
  >
    <Text>
      To use the indexing policy editor, please login to the
      <StyledLinkBase
        href="https://portal.azure.com"
        target="_blank"
      >
        azure portal.
      </StyledLinkBase>
    </Text>
  </StyledMessageBar>
  <Stack
    horizontal={true}
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Text
      styles={
        Object {
          "root": Object {
            "color": "windowtext",
            "fontSize": 14,
          },
        }
      }
    >
      Refreshing index transformation progress
    </Text>
    <StyledSpinnerBase
      size={1}
    />
  </Stack>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    You can make more indexing changes once the current index transformation is complete. 
    <StyledLinkBase
      onClick={[Function]}
    >
      Refresh to check if it has completed.
    </StyledLinkBase>
  </Text>
  <Text
    styles={
      Object {
        "root": Object {
          "color": "windowtext",
          "fontSize": 14,
        },
      }
    }
  >
    You can make more indexing changes once the current index transformation has completed. It is 90% complete. 
    <StyledLinkBase
      onClick={[Function]}
    >
      Refresh to check the progress.
    </StyledLinkBase>
  </Text>
</Fragment>
`;
